### PR TITLE
revise user domain

### PR DESCRIPTION
**Pull request by ro0NL**
flavor `User` entity as you like with `email / email+password / nickname` etc.

- built in user has no credentials but does implement `CredentialInterface`
- redesign command message; `new CreateUserCommand($email, $password) => ['email' => $email, 'password' => $password]` 
- update `Security` infra as needed
- move `User\PasswordEncoderInterface` to `User\Credential\Password\...`

### CODE REVIEW

File: src/User/Credential/EmailCredentialTrait.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * @author Roland Franssen <franssen.roland@gmail.com>
+ */
+trait EmailCredentialTrait
+{
+    private $email;
+
+    public function getEmail(): string
+    {
+        return $this->email;
```
Comment (iltar) [inline src/User/Credential/EmailCredentialTrait.php]: This trait isn't very useful if you want to have a custom `getEmail` I think. Could make this final and simply use `$this->email` in `getUsername`. The same goes for the other traits.

Comment (ro0NL) [reply src/User/Credential/EmailCredentialTrait.php]: :+1: updated.

File: src/User/Password/PasswordAlgorithm.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+        return $instance;
+    }
+
+    public function __toString(): string
+    {
+        return (string) $this->type;
+    }
+
+    private function __construct()
```
Comment (iltar) [inline src/User/Password/PasswordAlgorithm.php]: Seems like `$type` and `$legacy` are mandatory parameters, could be moved to the constructor to reflect this

File: src/User/Password/PasswordAlgorithm.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    /** @var array|null */
+    public $options;
+
+    /** @var string|null */
+    public $salt;
+
+    /** @var bool */
+    public $legacy;
+
+    public static function create(int $type = \PASSWORD_DEFAULT, array $options = []): self
```
Comment (iltar) [inline src/User/Password/PasswordAlgorithm.php]: The create methods should probably receive some documentation explaining the values. There's currently no way of telling what `$options` can contain or what the difference between int and string `$type` is.

File: src/User/Password/PasswordAlgorithm.php (hunk 2 above)
Comment (ro0NL) [reply src/User/Password/PasswordAlgorithm.php]: Now less LOC!

File: src/User/Password/PasswordAlgorithm.php (hunk 3 above)
Comment (ro0NL) [reply src/User/Password/PasswordAlgorithm.php]: should be good now

**Comment by ro0NL** [comment]
base design ready :)

**Pull request merged**